### PR TITLE
drivers/periph: Add documentation on thread safety and initialization

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

As the title says.

### Testing procedure

The CI will generate a doc preview that can be reviewed.

### Issues/PRs references

None

### CODE REVIEW

File: drivers/include/periph/doc.txt [hunk 1]
```diff
@@ -16,6 +16,64 @@
...
  *
+ * # Initialization and Thread Safety
+ *
+ * Peripheral drivers fall into three categories regarding to their
+ * initialization:
+ *
+ * 1. Drivers that do not need to be initialized
+ *    (e.g. @ref drivers_periph_eeprom)
+ * 2. Drivers that are automatically initialized by the module `periph_init`,
+ *    (see @ref drivers_periph_init for details)
```
Comment (bergzand) [inline drivers/include/periph/doc.txt:27]:
```suggestion
 * 2. Drivers that are automatically initialized by the module `periph_init` at boot,
 *    (see @ref drivers_periph_init for details)
```

In the paragraph below it mentions drivers initialized at boot. Better make it explicit which category this is.

File: drivers/include/periph/doc.txt [hunk 2]
```diff
@@ -16,6 +16,64 @@
  * defines a standardized interface to access MCU peripherals that is not tied
  * to any specific vendor, platform or architecture.
  *
+ * # Initialization and Thread Safety
```
Comment (chrysn) [inline drivers/include/periph/doc.txt:19]: Thread gets turned into a hyperlink thanks to automagic and mqtt paho.

Not sure whether there is anything we can do here -- escaping the word is probably a whack-a-mole...

File: drivers/include/periph/doc.txt [hunk 3]
```diff
@@ -16,6 +16,64 @@
...
+ *
+ * | Peripheral Driver  | Initialized By    | Thread Safety                                                                 |
+ * |:------------------ |:----------------- |:----------------------------------------------------------------------------- |
+ * | `periph_adc`       | user / driver     | Partial (no concurrent use of the same ADC line allowed)                      |
+ * | `periph_can`       | user / driver     | Partial (no concurrent use of the same CAN device allowed)                    |
+ * | `periph_cpuid`     | not needed        | Full Thread Safety                                                            |
+ * | `periph_dac`       | user / driver     | Partial (no concurrent use of the same DAC line allowed)                      |
+ * | `periph_eeprom`    | not needed        | None (no concurrency whatsoever)                                              |
+ * | `periph_flashpage` | not needed        | None (no concurrency whatsoever)                                              |
+ * | `periph_gpio`      | not needed        | Limited (reads are fine, concurrent writes to pins on distinct ports work)    |
```
Comment (chrysn) [inline drivers/include/periph/doc.txt:55]:
```suggestion
 * | `periph_gpio`      | not needed        | Limited (reads are fine, concurrent writes to pins are allowed on distinct ports)    |
```

A lot of things work that are not allowed.

**Review by chrysn — APPROVED** [review]
Great addition.

Deep links:
* https://ci.riot-os.org/results/db88ef36008c4cefa0e8ba6687e2fa9a/doc-preview/group__drivers__periph.html
* https://ci.riot-os.org/results/db88ef36008c4cefa0e8ba6687e2fa9a/doc-preview/group__drivers__periph__init.html#ga2cd4bdd061501508fe71aa672f690e81

**Comment by benpicco** [comment]
bors merge

**Comment by aabadie** [comment]
Please squash!

**Comment by aabadie** [comment]
bors merge

**Comment by aabadie** [comment]
bors cancel

**Comment by aabadie** [comment]
bors merge

**Pull request merged**